nixos/modules/services/cluster/k3s/default.nix

```
@@ -190,5 +190,25 @@ in
         );
       };
     };
+
+    # K3s doesn't stop containers when service is stopped.
+    # On host shutdown, containers are still running and needs to be stopped for avoiding data corruption.
+    systemd.services.killall-k3s = {
+      # "killall-k3s" is used as service name, because naming `k3s-*` introduces a loop at `k3s-killall.sh`, delaying shutdown.
+      description = "Executes k3s-killall.sh on shutdown";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        User = "root";
```

```suggestion
        User = "root";
```
I have been using this routine as `root`. Then I wonder, if some other user level would have enough permission for making it work.